File: C++/541.cpp
__________________________________________________________________________________________________
sample 4 ms submission
static int x=[](){
    std::ios::sync_with_stdio(false);
    cin.tie(NULL);
    return 0;
}();

class Solution {
public:
    string reverseStr(string s, int k) {
        if (k == 0) return s;
        int i = 0;
        while (i < (int) s.size() - k + 1) {
            for (int j = 0; j < k / 2; j++) {
                swap(s[i + j], s[i + k - j - 1]);
            }
            i += (2 * k);
        }
        // i -= (2 * k);
        int len = (int) s.size() - i;
        if (len < k) {
            for (int j = 0; j < len / 2; j++) {
                swap(s[i + j], s[s.size() - j - 1]);
            }
        }
        return s;
    }
};
__________________________________________________________________________________________________
sample 9396 kb submission
class Solution {
public:
    string reverseStr(string s, int k) {
        for (int i = 0; i < s.size(); i += 2 * k)
            reverse(s, i, min(k, (int) s.size() - i));
        
        return s;
    }
    
    void reverse(string &s, int begin, int length) {
        for (int i = 0; i < length / 2; ++i)
            swap(s[begin + i], s[begin + length - i - 1]);
    }
};

static auto _ = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    return nullptr;
}();
__________________________________________________________________________________________________

File: Java/541.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public String reverseStr(String s, int k) {
        char[] res = s.toCharArray();
        for(int i = 0; i < s.length(); i+=2*k){
            swap(res,i,Math.min(i+k-1,s.length()-1));
        }
        return new String(res);
    }
    
    private void swap(char[] res, int start, int end){
        while(start < end){
            char tmp = res[start];
            res[start++]= res[end];
            res[end--]= tmp;
        }
    }
}
__________________________________________________________________________________________________
sample 37388 kb submission
class Solution {
    public String reverseStr(String s, int k) {
       char[] chs = s.toCharArray();
       int len = chs.length;        
       boolean reverseFlag = true;     
       for(int i = 0; i < len; i += k){
           if(reverseFlag){
               int left = i;
               int right = Math.min(left + k - 1, len - 1);
               reversePartOfCharArray(chs, i, right);
           }
           reverseFlag = !reverseFlag;
       } 
         
       return String.valueOf(chs);
    }
    
    public void reversePartOfCharArray(char[] chs, int l, int r){
        while(l < r){
            char tmp = chs[l];
            chs[l] = chs[r];
            chs[r] = tmp;
            l ++;
            r --;
        }
    }
}
__________________________________________________________________________________________________
